### PR TITLE
Report the output of arbitrary commands.

### PR DESCRIPTION
**Pull request by UmanShahzad**
Right now, we can't glean any useful information from running an arbitrary command, except if it passed or not.

This PR makes arbitrary sanity check commands log their full output to the sanity check report, so we know what's actually going on.

To test, install the requirements into a virtualenv, cd to the tests directory and run

    python -m unittest test_sanity_check

### CODE REVIEW

File: files/sanity_check.py [hunk 1]
```diff
@@ -49,13 +49,10 @@ def check_port(host, port, retries=DEFAULT_PORT_RETRIES, retry_delay=DEFAULT_POR
...
 def check_command(command):
     try:
-        devnull = open(os.devnull, 'w')
-        subprocess.check_call(
-            command, shell=True, stdout=devnull, stderr=subprocess.STDOUT
-        )
-        return True
-    except subprocess.CalledProcessError:
-        return False
+        output = subprocess.check_output(command, shell=True, stderr=subprocess.STDOUT)
```
Comment (e-kolpakov) [inline files/sanity_check.py:52]: @UmanShahzad Might make sense to rename to `run_command` or something similar - it's not just checking if command execution succeeded, but also return output.

File: files/sanity_check.py [hunk 2]
```diff
@@ -129,13 +126,12 @@ def test_ports(ports):
...
-        if not check_command(command['command']):
-            response.append(SanityCheckResult(False, command['message']))
-        else:
-            message = format("Command {} returned exit status 0".format(
-                command['command']
-            ))
-            response.append(SanityCheckResult(True, message))
+        command_passed, command_output = check_command(command['command'])
+        command_message = (
+            "Command {} returned exit status 0".format(command['command'])
```
Comment (e-kolpakov) [inline files/sanity_check.py]: @UmanShahzad if command failed most likely (or absolutely always?) it returns non-zero status code. This message will be quite misleading.

And in general, it would look like:

    * "Command bash returned exit code 0" if `command_passed` is true
    * "Command Exception: something very bad happened + giant stacktrace .... snip .... few dozen lines later returned exit status 0" if `command_passed` is false.

Maybe something like

```
if command_passed:
    command_message = "Command {} returned status code 0"
else:
    command_message=  "Command {} returned status code {}; stdout/stderr was {}".format(...)
```

File: tests/test_sanity_check.py [hunk 3]
```diff
@@ -85,22 +85,34 @@ def test_check_port_retries_negative(self):
...
-        self.assertFalse(sanity_check.check_command("exit 1 && echo 1"))
+        command_passed, command_output = sanity_check.check_command("exit 1 && echo 1")
+        self.assertFalse(command_passed)
+        self.assertEqual(command_output, '')
 
     def test_command_negative_pipe(self):
-        self.assertFalse(sanity_check.check_command("echo foo bar | grep canary"))
+        command_passed, command_output = sanity_check.check_command("echo foo bar | grep canary")
+        self.assertFalse(command_passed)
+        self.assertEqual(command_output, '')
```
Comment (e-kolpakov) [inline tests/test_sanity_check.py]: @UmanShahzad Might be good to add a test scenario with exist code != 0 and some stdout/stderr output.

File: tests/test_sanity_check.py [hunk 4]
```diff
@@ -216,6 +228,7 @@ def test_global_free_space_percentage(self):
         )
         self.assertEqual(report, responses)
 
+
```
Comment (e-kolpakov) [inline tests/test_sanity_check.py:236]: @UmanShahzad Might be unnecessary change (unless it was pep8 complaining about something)

**Review by e-kolpakov — CHANGES_REQUESTED** [review]
@UmanShahzad looks good, except message in case of failing command might be quite misleading + there are no test for non-zero exit code and stdout/stderr output.

File: files/sanity_check.py (hunk 2 above)
Comment (UmanShahzad) [reply files/sanity_check.py]: @e-kolpakov I'm confused what you mean by this. You may have mistaken how the Python code executes here -- it only uses the "...exit status 0" string if it really is exit status 0, i.e. if the command passed. Otherwise it makes `command_message` equal to the configured failure message plus whatever the command outputted.

The outputs look like this in the report:

Passed:
```
#. OK Command rabbitmqctl node_health_check returned exit status 0
Timeout: 70.0 seconds
Checking health of node rabbit@rabbitmq
Health check passed
#. OK Enough free space /. Free space is 85.0%
#. OK Port localhost:5671 is getting connections
#. OK Port localhost:15671 is getting connections
```

Not passed:
```
#. ERROR RabbitMQ health check not passing
Timeout: 70.0 seconds
Checking health of node rabbit@rabbitmq
Error: {killed,{gen_server2,call,[<10994.27345.1912>,{info,[pid]},infinity]}}
#. OK Enough free space /. Free space is 85.0%
#. OK Port localhost:5671 is getting connections
#. OK Port localhost:15671 is getting connections
```

Also, we use the `command['message']` value as an error string, so "RabbitMQ health check not passing" comes in that as we set it [here](https://github.com/open-craft/ansible-playbooks/blob/master/group_vars/rabbitmq/public.yml#L15), and we want to use that for error cases; if not, it should be erased throughout all of ansible-playbooks -- definitely out of scope and probably not desired.

File: tests/test_sanity_check.py (hunk 3 above)
Comment (UmanShahzad) [reply tests/test_sanity_check.py]: @e-kolpakov If the command did not pass, that means it exited with code != 0. You can confirm this by reading the documentation about the `subprocess.check_output` function:

```
    If the exit code was non-zero it raises a CalledProcessError.  The
    CalledProcessError object will have the return code in the returncode
    attribute and output in the output attribute.
```

File: tests/test_sanity_check.py (hunk 4 above)
Comment (UmanShahzad) [reply tests/test_sanity_check.py:236]: @e-kolpakov Unfortunately we don't have pep8 (or any quality check) on this repository's Python code, but this is definitely something it'd complain about. I don't want to expand this PR so much as to include pep8 whole, but I noticed this so added it -- leaving the code just a bit better than I found it, which is preferred over making whole separate PRs.

File: files/sanity_check.py (hunk 1 above)
Comment (UmanShahzad) [reply files/sanity_check.py:52]: @e-kolpakov You mean the `check_command` function is to change, right? I'll confirm first before making changes, but that sounds good to me to change.

Comment (e-kolpakov) [reply files/sanity_check.py:52]: @UmanShahzad yes `check_command` - > `run_command`

File: files/sanity_check.py (hunk 2 above)
Comment (e-kolpakov) [reply files/sanity_check.py]: @UmanShahzad right, now I see it. This code is a bit confusing though - maybe something more traditional would be less confusing? Like

```
if success:
    message = "Command {} returned 0"
else:
    message = command['message'] 
```

File: tests/test_sanity_check.py (hunk 3 above)
Comment (e-kolpakov) [reply tests/test_sanity_check.py]: @UmanShahzad that's not exactly what I meant. My point was that as far as I can tell, there are no test with `command_passed = false` and `command_output != ''`. Might make sense to add it to ensure that command stdout/stderr is actually captured and returned (especially stderr, since it doesn't seem to be covered by other existing tests)

**Review by e-kolpakov — APPROVED** [review]
@UmanShahzad looks good, but a few improvement suggestions.

File: files/sanity_check.py (hunk 2 above)
Comment (UmanShahzad) [reply files/sanity_check.py]: This is actually idiomatic in Python but since it did throw someone off, it might happen again so OK.

File: files/sanity_check.py (hunk 1 above)
Comment (UmanShahzad) [reply files/sanity_check.py:52]: :+1:

File: tests/test_sanity_check.py (hunk 3 above)
Comment (UmanShahzad) [reply tests/test_sanity_check.py]: Ah, ok! I'll add a test for this.

**Comment by UmanShahzad** [comment]
@e-kolpakov Thanks!

Addressed review with 2cd32a00108ed65b9f86fd130e890821af5f3a2a. (Ignore the `"""`s, I deleted those later).
Squashing and merging now.

File: files/sanity_check.py (hunk 2 above)
Comment (e-kolpakov) [reply files/sanity_check.py]: It is a bit less idiomatic since it is put onto different lines.

Ternary if is somewhat confusion-prone in any language, and in python it's even more since you need to read it from the middle.

I mean, this is ok(ish) 

```
x = 42 if true else 96
```

But multi-line ternary if is not quite :)

Comment (UmanShahzad) [reply files/sanity_check.py]: :+1: makes sense.
